File: scripts/check_version.py
from __future__ import annotations

import re
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]


def extract(pattern: str, path: Path) -> str:
    match = re.search(pattern, path.read_text(encoding="utf-8"), re.MULTILINE)
    if match is None:
        raise SystemExit(f"version not found in {path.name}")
    return match.group(1)


versions = {
    "package": extract(r'^__version__ = "([^"]+)"$', ROOT / "src/hufu/__init__.py"),
    "pyproject": extract(r'^version = "([^"]+)"$', ROOT / "pyproject.toml"),
    "version_doc": extract(r"Current version: `([^`]+)`", ROOT / "VERSION.md"),
}

if len(set(versions.values())) != 1:
    raise SystemExit(f"version mismatch: {versions}")

print(f"version check passed: {next(iter(versions.values()))}")

File: scripts/check-version.mjs
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const root = join(dirname(fileURLToPath(import.meta.url)), "..");

function extract(pattern, filePath, label) {
  const text = readFileSync(filePath, "utf8");
  const match = text.match(pattern);
  if (match === null) {
    console.error(`version not found in ${label}`);
    process.exit(1);
  }
  return match[1];
}

const packageJson = JSON.parse(readFileSync(join(root, "package.json"), "utf8"));
const pluginPackage = JSON.parse(
  readFileSync(join(root, "packages/hufu-dsh/package.json"), "utf8"),
);
const versions = {
  package: packageJson.version,
  plugin: pluginPackage.version,
  source: extract(
    /^export const VERSION = "([^"]+)";$/m,
    join(root, "src/hufu/version.ts"),
    "version.ts",
  ),
  versionDoc: extract(
    /Current version: `([^`]+)`/,
    join(root, "VERSION.md"),
    "VERSION.md",
  ),
};

if (new Set(Object.values(versions)).size !== 1) {
  console.error(`version mismatch: ${JSON.stringify(versions)}`);
  process.exit(1);
}

console.log(`version check passed: ${versions.package}`);
